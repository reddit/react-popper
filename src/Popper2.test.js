// @flow
import React from 'react';
import { mount } from 'enzyme';

// Private API
import { InnerPopper } from './Popper2';

const mountPopper = props =>
  mount(
    <InnerPopper {...props}>
      {({ ref, style, placement, arrowProps }) => (
        <div ref={ref} style={style} data-placement={placement}>
          <div {...arrowProps} />
        </div>
      )}
    </InnerPopper>
  );

describe('Popper component', () => {
  it('renders the expected markup', () => {
    const referenceElement = document.createElement('div');
    const wrapper = mountPopper({ referenceElement });
    expect(wrapper).toMatchSnapshot();
  });

  it('initializes the Popper.js instance on first update', () => {
    const referenceElement = document.createElement('div');
    const wrapper = mountPopper({ referenceElement });
    expect(wrapper.instance().popperInstance).toBeDefined();
  });

  it("doesn't update Popper.js instance on props update if not needed by Popper.js", () => {
    const referenceElement = document.createElement('div');
    const wrapper = mountPopper({ referenceElement, placement: 'bottom' });
    const instance = wrapper.instance().popperInstance;

    expect(instance).toBeDefined();

    wrapper.setProps({ placement: 'bottom' });

    expect(wrapper.instance().popperInstance).toBe(instance);
  });

  it('updates Popper.js on explicitly listed props change', () => {
    const referenceElement = document.createElement('div');
    const wrapper = mountPopper({ referenceElement });
    const instance = wrapper.instance().popperInstance;
    wrapper.setProps({ placement: 'top' });
    wrapper.update();
    expect(wrapper.instance().popperInstance).not.toBe(instance);
  });

  it('does not update Popper.js on generic props change', () => {
    const referenceElement = document.createElement('div');
    const wrapper = mountPopper({ referenceElement });
    const instance = wrapper.instance().popperInstance;
    wrapper.setProps({ foo: 'bar' });
    wrapper.update();
    expect(wrapper.instance().popperInstance).toBe(instance);
  });

  it('destroys Popper.js instance on unmount', () => {
    const referenceElement = document.createElement('div');
    const wrapper = mountPopper({ referenceElement });
    const instance = wrapper.instance().popperInstance;
    wrapper.unmount();
    expect(instance.state.isDestroyed).toBe(true);
  });

  it('handles changing refs gracefully', () => {
    const referenceElement = document.createElement('div');
    expect(() =>
      mount(
        <InnerPopper referenceElemen={referenceElement}>
          {({ ref, style, placement, arrowProps }) => (
            <div
              ref={current => ref(current)}
              style={style}
              data-placement={placement}
            >
              <div {...arrowProps} ref={current => arrowProps.ref(current)} />
            </div>
          )}
        </InnerPopper>
      )
    ).not.toThrow();
  });

  it('accepts a `referenceElement` property', () => {
    class VirtualReference {
      getBoundingClientRect() {
        return {
          top: 10,
          left: 10,
          bottom: 20,
          right: 100,
          width: 90,
          height: 10,
        };
      }

      get clientWidth() {
        return this.getBoundingClientRect().width;
      }

      get clientHeight() {
        return this.getBoundingClientRect().height;
      }
    }

    const virtualReferenceElement = new VirtualReference();
    const wrapper = mountPopper({ referenceElement: virtualReferenceElement });

    expect(wrapper.instance().popperInstance.reference).toBe(
      virtualReferenceElement
    );
  });
});
