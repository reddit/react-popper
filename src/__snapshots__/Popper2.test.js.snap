// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper component renders the expected markup 1`] = `
<InnerPopper
  eventsEnabled={true}
  placement="bottom"
  positionFixed={false}
  referenceElement={<div />}
>
  <div
    style={
      Object {
        "left": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
      }
    }
  >
    <div
      style={Object {}}
    />
  </div>
</InnerPopper>
`;
